src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,164 @@
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            Node<T> item = new Node<>(null, value, null);
+            head = item;
+            tail = item;
+        } else {
+            Node<T> item = new Node<T>(tail, value, null);
+            tail.next = item;
+            tail = item;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        inSize(index);
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> second = head;
+            head = new Node<>(null, value, second);
+            second.prev = head;
+            size++;
+            return;
+        }
+        Node<T> current = findNodeByIndex(index);
+        Node<T> newItem = new Node<>(null, value, null);
+        newItem.next = current;
+        newItem.prev = current.prev;
+        current.prev.next = newItem;
+        current.prev = newItem;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> current = findNodeByIndex(index);
+        T result = current.item;
+        current.item = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        inSize(index);
+        if (size == 1) {
+            final T removed = head.item;
+            head = null;
+            tail = null;
+            size--;
+            return removed;
+        }
+        if (index == 0) {
+            head = head.next;
+            size--;
+            return head.prev.item;
+        } else if (index == size - 1) {
+            tail = tail.prev;
+            size--;
+            return tail.next.item;
+        }
+        Node<T> current = findNodeByIndex(index);
+        return unlink(current);
     }
 
     @Override
     public boolean remove(T object) {
+        for (Node<T> i = head; i != null; i = i.next) {
+            if (i.item == object || object != null && object.equals(i.item)) {
+                unlink(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        inSizeWithSize(index);
+        int indexLinked = 0;
+        Node<T> current = head;
+        while (indexLinked != index) {
```

Comment:
let's do start from head if the index is in the first half of the list and from tail in another case